File: solutions/haskell/069-balanced-binary-tree.hs
{-
  Problem 69: Balanced Binary Tree (LeetCode 110)
  Difficulty: Easy
  Language: Haskell
-}
isBalanced :: Tree a -> Bool
isBalanced = (/= -1) . height
  where
    height Nil = 0
    height (Node _ l r) =
      let lh = height l; rh = height r
      in if lh == -1 || rh == -1 || abs (lh - rh) > 1 then -1
         else 1 + max lh rh

File: solutions/clojure/069-balanced-binary-tree.clj
;; Problem 69: Balanced Binary Tree (LeetCode 110)
;; Difficulty: Easy
;; Language: Clojure
;; 
(defn balanced? [root]
  (letfn [(height [node]
            (if (nil? node) 0
              (let [l (height (:left node))
                    r (height (:right node))]
                (if (or (= l -1) (= r -1) (> (abs (- l r)) 1)) -1
                  (inc (max l r))))))]
    (not= (height root) -1)))
